home_work_9/index.js

```diff
@@ -0,0 +1,173 @@
+const validateValue = (value) => {
+  if (value === undefined || value === null) {
+    return;
+  }
+  return value;
+};
+
+const EMPTY_VALUE_MESSAGE = "This value is not specified";
+
+function addSpaseInArray(array) {
+  if (validateValue(array) || !array?.length === 0)
+    return array.map((el) => ` ${el}`);
+  else {
+    return EMPTY_VALUE_MESSAGE;
+  }
+}
+
+const testObject1 = { key1: 2, key2: "blabla" };
+const testObject2 = { key1: 2, key2: "blabla" };
+
+//1.
+Function.prototype.myOwnBind = function (context, ...arg) {
+  const fn = this;
+  return function () {
+    fn.call(context, ...arg);
+  };
+};
+function showTestObject(name, age) {
+  console.log(this, name, age);
+}
+
+showTestObject.myOwnBind(testObject1, "test name", 23)();
+
+//2.
+function Man(name, age, sex, nationality, country, listCountriesToVisit) {
+  this.name = name;
+  this.age = age;
+  this.sex = sex;
+  this.nationality = nationality;
+  this.country = country;
+  this.listCountriesToVisit = listCountriesToVisit;
+}
+
+function introduce() {
+  if (this !== window) {
+    return `Hello! My name is ${this.name}. I am ${
+      this.age
+    } years old. Now I live in ${
+      this.country
+    }. I dream to visit:${addSpaseInArray(this.listCountriesToVisit)}`;
+  } else {
+    return EMPTY_VALUE_MESSAGE;
+  }
+}
+function showWakeUpMessage() {
+  return `Good morning ${this.country}! I am ${
+    this.nationality
+  } and I am determined to visit:${addSpaseInArray(this.listCountriesToVisit)}`;
+}
+function showGoToBedMessage() {
+  return `Good night! I'll go to bed. I hope I'll dream of my beloved ${this.country}`;
+}
+function showListCountriesToVisit() {
+  return `Countries ${this.name} from ${
+    this.country
+  } wants to visit:${addSpaseInArray(this.listCountriesToVisit)}`;
+}
+
+const Den = new Man("Den", 25, "male", "Ukrainian", "Ukraine", [
+  "Poland",
+  "Portugal",
+  "France",
+]);
+console.log(showWakeUpMessage.apply(Den));
+console.log(showListCountriesToVisit.apply(Den));
+console.log(showGoToBedMessage.call(Den));
+console.log(introduce.apply(Den));
+
+//3.
+function checkObjectsEqual(object1, object2) {
```

Comment:
This one is good for flat objects, but what if we will have nested?

<img width="540" alt="image" src="https://user-images.githubusercontent.com/84010501/194695585-2771d5d5-aead-4e2f-83cd-0d36934c3e8d.png">